# Анализ базы данных сервиса для чтения книг

# Описание исследования

Компания купила крупный сервис для чтения книг по подписке. В ней находится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Цель исследования
 
Проанализировать базу данных, сделать выводы и дать рекомендации для нового сервиса по чтению книг.

# Задачи исследования

1. Подключиться к базе данных.
2. Оценить представленные данные.
3. Выполнить соответствующие запросы.
4. Написать выводы по итогу выполнения каждой из запроса.

# Исходные данные

Таблица books содержит данные о книгах:

    book_id — идентификатор книги;
    author_id — идентификатор автора;
    title — название книги;
    num_pages — количество страниц;
    publication_date — дата публикации книги;
    publisher_id — идентификатор издателя.

Таблица authors содержит данные об авторах:

    author_id — идентификатор автора;
    author — имя автора.

Таблица publishers содержит данные об издательствах:

    publisher_id — идентификатор издательства;
    publisher — название издательства;

Таблица ratings содержит данные о пользовательских оценках книг:

    rating_id — идентификатор оценки;
    book_id — идентификатор книги;
    username — имя пользователя, оставившего оценку;
    rating — оценка книги.

Таблица reviews содержит данные о пользовательских обзорах:

    review_id — идентификатор обзора;
    book_id — идентификатор книги;
    username — имя автора обзора;
    text — текст обзора.

## Подключение к базе данных

### Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine
from PIL import Image
from io import BytesIO
from IPython.display import Image

### Устанавливаем параметры

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

### Сохраняем коннектор

In [3]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Оценка представленных данных

### Отобразим схему базы данных для исследования

In [4]:
Image(url='https://pictures.s3.yandex.net/resources/scheme_1589269096.png', width=1000, height=600)

### Выведем первые 5 строк таблицы books

In [5]:
query = '''SELECT *
           FROM books
           LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


### Создадим функцию database_table_output, которая будет выводить первые 5 строк в зависимости от изучаемой таблицы базы данных

In [6]:
def database_table_output(table_name):
    query = f'''SELECT *
                FROM {table_name}
                LIMIT 5'''
    con = engine.connect()
    return pd.io.sql.read_sql(sql=text(query), con=con)

### Посчитаем количество строк в таблице books

In [7]:
query = '''SELECT COUNT(*)
           FROM books'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,count
0,1000


### Создадим функцию count_table_output, которая будет количество строк в изучаемой таблице

In [8]:
def count_table_output(table_name):
    query = f'''SELECT COUNT(*)
                FROM {table_name}'''
    con = engine.connect()
    return pd.io.sql.read_sql(sql=text(query), con=con)

### Выведем первые 5 строк таблицы authors, используя функцию database_table_output

In [9]:
database_table_output('authors')

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


### Посчитаем количество строк в таблице authors, используя функцию count_table_output

In [10]:
count_table_output('authors')

,count
0,636


### Выведем первые 5 строк таблицы publishers, используя функцию database_table_output

In [11]:
database_table_output('publishers')

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


### Посчитаем количество строк в таблице publishers, используя функцию count_table_output

In [12]:
count_table_output('publishers')

,count
0,340


### Выведем первые 5 строк таблицы ratings, используя функцию database_table_output

In [13]:
database_table_output('ratings')

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


### Посчитаем количество строк в таблице ratings, используя функцию count_table_output

In [14]:
count_table_output('ratings')

,count
0,6456


### Выведем первые 5 строк таблицы reviews, используя функцию database_table_output

In [15]:
database_table_output('reviews')

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


### Посчитаем количество строк в таблице reviews, используя функцию count_table_output

In [16]:
count_table_output('reviews')

,count
0,2793


### Выводы по оценке представленных данных

1. Таблица books содержит 6 столбцов и 1000 строк. В ней есть целые числа int, строки varchar, даты datetime и тип id.
2. Таблица authors содержит 2 столбца и 636 строк. В ней есть целые числа int и строки varchar.
3. Таблица publishers содержит 2 столбца и 340 строк. В ней есть целые числа int и строки varchar.
4. Таблица publishers содержит 4 столбца и 6456 строк. В ней есть целые числа int и строки varchar.
5. Таблица reviews содержит 4 столбца и 2793 строк. В ней есть целые числа int и строки varchar.

## Выполнение соответствующих запросов

### Посчитаем количество книг, которые вышли после 1 января 2000 года

In [17]:
query = ''' SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01' '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,count
0,819


### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [18]:
query = '''SELECT b.title,
                  COUNT(DISTINCT rev.review_id) as count_review,
                  AVG(rat.rating) as avg_rating
           FROM books as b
           LEFT OUTER JOIN reviews as rev ON b.book_id = rev.book_id
           LEFT OUTER JOIN ratings as rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           ORDER BY count_review DESC'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,title,count_review,avg_rating
0,Twilight (Twilight #1),7,3.662500
1,Water for Elephants,6,3.977273
2,The Glass Castle,6,4.206897
3,Harry Potter and the Prisoner of Azkaban (Harr...,6,4.414634
4,The Curious Incident of the Dog in the Night-Time,6,4.081081
...,...,...,...
995,Anne Rice's The Vampire Lestat: A Graphic Novel,0,3.666667
996,The Natural Way to Draw,0,3.000000
997,The Cat in the Hat and Other Dr. Seuss Favorites,0,5.000000
998,Essential Tales and Poems,0,4.000000


### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [19]:
query = '''SELECT p.publisher
           FROM publishers as p
           LEFT OUTER JOIN books as b ON p.publisher_id = b.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher
           ORDER BY COUNT(b.book_id) DESC
           LIMIT 1'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,publisher
0,Penguin Books


### Определим автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками

In [20]:
query = '''SELECT author,
                  AVG(avg_rating) as average_rating
           FROM(SELECT a.author,
                       b.author_id,
                       AVG(rat.rating) as avg_rating
                FROM books as b
                LEFT OUTER JOIN ratings as rat ON b.book_id = rat.book_id
                LEFT OUTER JOIN authors as a ON a.author_id = b.author_id
                GROUP BY b.author_id, b.book_id, a.author
                HAVING COUNT(rat.rating_id) >= 50
                ORDER BY avg_rating DESC) as avg_data
           GROUP BY author
           ORDER BY average_rating DESC
           LIMIT 1'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,author,average_rating
0,J.K. Rowling/Mary GrandPré,4.283844


### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [21]:
query = '''SELECT AVG(reviews_count) as avg_reviews
           FROM (SELECT DISTINCT rev.username,
                        COUNT(DISTINCT rev.review_id) AS reviews_count,
                        COUNT(DISTINCT rat.rating_id) AS ratings_count
                 FROM reviews as rev
                 LEFT OUTER JOIN ratings as rat ON rev.username = rat.username
                 GROUP BY rev.username
                 HAVING COUNT(DISTINCT rat.rating_id) > 48
                 ORDER BY ratings_count) as acive_users'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

,avg_reviews
0,24.0


## Выводы и рекомендации

### Выводы по исследовательскому анализу данных

1. После 1 января 2000 года вышло 819 книг.
2. Максимальное количество обзоров в количестве 7 штук имеется у книги "Twilight (Twilight #1)".
3. Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц.
4. Автором с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, является J.K. Rowling/Mary GrandPré со средней оценкой в 4.28.
5. Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24.

### Ценностное предложение для нового сервиса по чтению книг

Сделать рекомендательную систему книг читателям на основе популярности книги, рейтинга и количества обзоров. Например предлагать читателям книги автора J.K. Rowling/Mary GrandPré, так как это достаточно известный и популярный автор с высокой средней оценкой ее книг.